src/array_api_extra/_apply.py

```diff
@@ -0,0 +1,351 @@
+"""Public API Functions."""
+
+# https://github.com/scikit-learn/scikit-learn/pull/27910#issuecomment-2568023972
+from __future__ import annotations
+
+from collections.abc import Callable, Hashable, Mapping, Sequence
+from functools import wraps
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, cast
+
+if TYPE_CHECKING:
+    # https://github.com/scikit-learn/scikit-learn/pull/27910#issuecomment-2568023972
+    from typing import TypeAlias
+
+    import numpy as np
+    import numpy.typing as npt
+
+    NumpyObject: TypeAlias = npt.NDArray[Any] | np.generic  # type: ignore[no-any-explicit]
+
+from ._lib._compat import (
+    array_namespace,
+    is_dask_namespace,
+    is_jax_namespace,
+)
+from ._lib._typing import Array, DType
+
+
+def apply_numpy_func(  # type: ignore[no-any-explicit]
+    func: Callable[..., NumpyObject | Sequence[NumpyObject]],
+    *args: Array,
+    shapes: Sequence[tuple[int, ...]] | None = None,
+    dtypes: Sequence[DType] | None = None,
+    xp: ModuleType | None = None,
+    input_indices: Sequence[Sequence[Hashable]] | None = None,
+    core_indices: Sequence[Sequence[Hashable]] | None = None,
+    output_indices: Sequence[Sequence[Hashable]] | None = None,
+    adjust_chunks: Sequence[dict[Hashable, Callable[[int], int]]] | None = None,
+    new_axes: Sequence[dict[Hashable, int]] | None = None,
+    **kwargs: Any,
+) -> tuple[Array, ...]:
+    """
+    Apply a function that operates on NumPY arrays to any Array API compliant arrays.
+
+    Parameters
+    ----------
+    func : callable
+        The function to apply. It must accept one or more NumPy arrays or generics as
+        positional arguments and return either a single NumPy array or generic, or a
+        tuple or list thereof.
+
+        It must be a pure function, i.e. without side effects such as disk output,
+        as depending on the backend it may be executed more than once.
+    *args : Array
+        One or more Array API compliant arrays.
+    shapes : Sequence[tuple[int, ...]], optional
+        Sequence of output shapes, one for each output of `func`.
+        If `func` returns a single (non-sequence) output, this must be a sequence
+        with a single element.
```

Comment:
I tried overloading but it was a big headache when validating inputs. I found this approach much simpler.